### PR TITLE
feat(components): anchor component

### PR DESCRIPTION
**Pull request by Fuphoenixes**
Please make sure these boxes are checked before submitting your PR, thank you!

- [ ] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [ ] Make sure you are merging your commits to `dev` branch.
- [ ] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Comment by Fuphoenixes** [comment]
@element-plus/backers

**Comment by kooriookami** [comment]
Nice Component! Please add a version tag `2.6.0` here. 
image

**Comment by kooriookami** [comment]
@jw-foss Can you review the design of API?

File: docs/en-US/component/anchor.md [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+| type       | set Anchor type.                         | ^[enum]`'defalut' \| 'underline'`                     | `default` |
+| get-current-anchor  | customize the anchor highlight.   | ^[Function]`(currentHref: string) => string`          | -   |
+| direction     | Set Anchor direction.                 | ^[enum]`'vertical' \| 'horizontal'`                   | `vertical`   |
+
+### Anchor events
+
+| Name   | Description                    | Type                                   |
+| ------ | ------------------------------ | -------------------------------------- |
+| change | callback when the step changes | ^[Function]`(href: string) => void` |
+| link-click | Triggered when the user clicks on the link  | ^[Function]`(e: MouseEvent, href?: string) => void` |
```
Comment (kooriookami) [inline docs/en-US/component/anchor.md:106]:
```suggestion
```
I think use `click` is enough.

File: docs/en-US/component/anchor.md [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+| Property   | Description                              | Type                                                  | Default |
+| ---------- | ---------------------------------------- | ----------------------------------------------------- | ------- |
+| container  | scroll container.                        | `string` \| `HTMLElement` \| `Window `                | —       |
+| offset     | set the offset of the anchor scroll.     | `number`                                              | 0    |
+| bound      | the offset of the element starting to trigger the anchor. | `number`                             | 15      |
+| duration   | set the scroll duration of the container, in milliseconds.          | `number`                   | 300    |
+| marker     | Whether to show the marker.              | ^[boolean]                                            | true   |
+| type       | set Anchor type.                         | ^[enum]`'defalut' \| 'underline'`                     | `default` |
+| get-current-anchor  | customize the anchor highlight.   | ^[Function]`(currentHref: string) => string`          | -   |
```
Comment (kooriookami) [inline docs/en-US/component/anchor.md:98]:
```suggestion
| active | customize the anchor highlight.   | `string` \| ^[Function]`(currentHref: string) => string`          | -   |
```
I think `active` is short, and support the string type. Even I think we can remove this API, it seems useless.

File: docs/en-US/component/anchor.md [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+| ---------- | ---------------------------------------- | ----------------------------------------------------- | ------- |
+| title      | the text content of the anchor link.     | `string`                                              | —       |
+| href       | The address of the anchor link.          | `string`                                              | -    |
+
+### AnchorLink slots
+
+| Name    | Description |
+| ------- | ----------- |
+| default | the content of the anchor link.   |
+| sublist | slots for child links. |
```
Comment (kooriookami) [inline docs/en-US/component/anchor.md:134]:
```suggestion
| sub-link | slots for child links. |
```

File: docs/en-US/component/anchor.md [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+## Affix Mode
+
+Use the affix component to fix the anchor point within the page.
+
+:::demo
+
+anchor/affix
+
+:::
```
Comment (kooriookami) [inline docs/en-US/component/anchor.md:44]: Move the affix mode to the last position, avoid shading.

**Comment by kooriookami** [comment]
image

https://github.com/element-plus/element-plus/assets/38392315/9e4d4ffe-f438-404d-8095-0152e788ee6d

When I clicked the link, it jumped to other place.

**Comment by Fuphoenixes** [comment]
> image
> 
>  QQ2024131-112753-HD.mp4 
> When I clicked the link, it jumped to other place.

This is done by vitepress, normal pages will not do this.
[Reference here](https://github.com/vuejs/vitepress/blob/54042ef5760e0af5590b616d909e7f23effedc22/src/client/app/router.ts#L208C26-L208C30)

**Comment by kooriookami** [comment]
> > image
> > QQ2024131-112753-HD.mp4
> > When I clicked the link, it jumped to other place.
> 
> This is done by vitepress, normal pages will not do this. [Reference here](https://github.com/vuejs/vitepress/blob/54042ef5760e0af5590b616d909e7f23effedc22/src/client/app/router.ts#L208C26-L208C30)

OK~

File: docs/en-US/component/anchor.md (hunk 4 above)
Comment (Fuphoenixes) [reply docs/en-US/component/anchor.md]: done

File: docs/en-US/component/anchor.md (hunk 3 above)
Comment (Fuphoenixes) [reply docs/en-US/component/anchor.md]: done

File: docs/en-US/component/anchor.md (hunk 1 above)
Comment (Fuphoenixes) [reply docs/en-US/component/anchor.md]: done

File: docs/en-US/component/anchor.md (hunk 2 above)
Comment (Fuphoenixes) [reply docs/en-US/component/anchor.md:98]: This API might be useful. I saw that `antd` only added this after version 3.22.0.
image

File: docs/en-US/component/anchor.md [hunk 5]
```diff
@@ -131,4 +131,4 @@ anchor/underline
 | Name    | Description |
 | ------- | ----------- |
 | default | the content of the anchor link.   |
-| sublist | slots for child links. |
+| subLink | slots for child links. |
```
Comment (kooriookami) [inline docs/en-US/component/anchor.md:134]:
```suggestion
| sub-link | slots for child links. |
```

File: docs/examples/anchor/affix.vue [hunk 6]
```diff
@@ -3,10 +3,10 @@
     <el-anchor :offset="70" style="width: 300px">
       <el-anchor-link href="#basic-usage">Basic Usage</el-anchor-link>
       <el-anchor-link href="#horizontal-mode">Horizontal Mode</el-anchor-link>
-      <el-anchor-link href="#affix-mode">Affix Mode</el-anchor-link>
+      <el-anchor-link href="#scroll-container">Scroll Container</el-anchor-link>
       <el-anchor-link href="#anchor-api">
         Anchor API
-        <template #sublist>
+        <template #subLink>
```
Comment (kooriookami) [inline docs/examples/anchor/affix.vue:9]:
```suggestion
        <template #sub-link>
```

File: docs/examples/anchor/basic.vue [hunk 7]
```diff
@@ -2,10 +2,10 @@
   <el-anchor :offset="70">
     <el-anchor-link href="#basic-usage">Basic Usage</el-anchor-link>
     <el-anchor-link href="#horizontal-mode">Horizontal Mode</el-anchor-link>
-    <el-anchor-link href="#affix-mode">Affix Mode</el-anchor-link>
+    <el-anchor-link href="#scroll-container">Scroll Container</el-anchor-link>
     <el-anchor-link href="#anchor-api">
       Anchor API
-      <template #sublist>
+      <template #subLink>
```
Comment (kooriookami) [inline docs/examples/anchor/basic.vue:8]:
```suggestion
      <template #sub-link>
```

File: docs/examples/anchor/change.vue [hunk 8]
```diff
@@ -2,10 +2,10 @@
   <el-anchor :offset="70" @change="handleChange">
     <el-anchor-link href="#basic-usage">Basic Usage</el-anchor-link>
     <el-anchor-link href="#horizontal-mode">Horizontal Mode</el-anchor-link>
-    <el-anchor-link href="#affix-mode">Affix Mode</el-anchor-link>
+    <el-anchor-link href="#scroll-container">Scroll Container</el-anchor-link>
     <el-anchor-link href="#anchor-api">
       Anchor API
-      <template #sublist>
+      <template #subLink>
```
Comment (kooriookami) [inline docs/examples/anchor/change.vue:8]:
```suggestion
      <template #sub-link>
```

File: docs/examples/anchor/getCurrentAnchor.vue [hunk 9]
```diff
@@ -2,10 +2,10 @@
   <el-anchor :offset="70" :get-current-anchor="getCurrentAnchor">
     <el-anchor-link href="#basic-usage">Basic Usage</el-anchor-link>
     <el-anchor-link href="#horizontal-mode">Horizontal Mode</el-anchor-link>
-    <el-anchor-link href="#affix-mode">Affix Mode</el-anchor-link>
+    <el-anchor-link href="#scroll-container">Scroll Container</el-anchor-link>
     <el-anchor-link href="#anchor-api">
       Anchor API
-      <template #sublist>
+      <template #subLink>
```
Comment (kooriookami) [inline docs/examples/anchor/getCurrentAnchor.vue:8]:
```suggestion
      <template #sub-link>
```

File: docs/examples/anchor/underline.vue [hunk 10]
```diff
@@ -2,10 +2,10 @@
   <el-anchor type="underline" :offset="70">
     <el-anchor-link href="#basic-usage">Basic Usage</el-anchor-link>
     <el-anchor-link href="#horizontal-mode">Horizontal Mode</el-anchor-link>
-    <el-anchor-link href="#affix-mode">Affix Mode</el-anchor-link>
+    <el-anchor-link href="#scroll-container">Scroll Container</el-anchor-link>
     <el-anchor-link href="#anchor-api">
       Anchor API
-      <template #sublist>
+      <template #subLink>
```
Comment (kooriookami) [inline docs/examples/anchor/underline.vue:8]:
```suggestion
      <template #sub-link>
```

File: packages/components/anchor/src/anchor-link.vue [hunk 11]
```diff
@@ -4,10 +4,10 @@
       <slot>{{ title }}</slot>
     </a>
     <div
-      v-if="$slots.sublist && direction === 'vertical'"
+      v-if="$slots.subLink && direction === 'vertical'"
       :class="ns.e('list')"
     >
-      <slot name="sublist" />
+      <slot name="subLink" />
```
Comment (kooriookami) [inline packages/components/anchor/src/anchor-link.vue:10]:
```suggestion
      <slot name="sub-link" />
```

**Review by kooriookami — CHANGES_REQUESTED** [review]

File: packages/components/anchor/src/anchor-link.vue (hunk 11 above)
Comment (Fuphoenixes) [reply packages/components/anchor/src/anchor-link.vue]: done

File: docs/en-US/component/anchor.md (hunk 2 above)
Comment (kooriookami) [reply docs/en-US/component/anchor.md:98]: Can we change this API to `highlight` and support both string and function types?

Comment (kooriookami) [reply docs/en-US/component/anchor.md:98]: I searched the PR and Issues of antd, `get-current-anchor` is a default active link.
Maybe we can make it like `v-model`? If you have some ideas, you can contact me on WeChat when you are free.

File: docs/en-US/component/anchor.md [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+
+### Anchor Attributes
+
+| Property   | Description                              | Type                                                  | Default |
+| ---------- | ---------------------------------------- | ----------------------------------------------------- | ------- |
+| container  | scroll container.                        | `string` \| `HTMLElement` \| `Window `                | —       |
+| offset     | set the offset of the anchor scroll.     | `number`                                              | 0    |
+| bound      | the offset of the element starting to trigger the anchor. | `number`                             | 15      |
+| duration   | set the scroll duration of the container, in milliseconds.          | `number`                   | 300    |
+| marker     | Whether to show the marker.              | ^[boolean]                                            | true   |
```
Comment (kooriookami) [inline docs/en-US/component/anchor.md:86]:
```suggestion
| marker     | whether to show the marker              | ^[boolean]                                            | true   |
```
We don't need uppercase letters or period in description, others too.

File: packages/components/anchor/src/helper.ts [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+export function easeInOutCubic(t: number, b: number, c: number, d: number) {
+  const cc = c - b
+  t /= d / 2
+  if (t < 1) {
+    return (cc / 2) * t * t * t + b
+  }
+  return (cc / 2) * ((t -= 2) * t * t + 2) + b
+}
```
Comment (kooriookami) [inline packages/components/anchor/src/helper.ts:113]: We can use https://vueuse.org/core/useTransition/
May be just use useScroll or useWindowScroll is better.

File: packages/components/anchor/src/helper.ts [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+      from,
+      to,
+      duration
+    )
+
+    if (isWindow(container)) {
+      container.scrollTo(window.pageXOffset, nextScrollTop)
+    } else {
+      container.scrollTop = nextScrollTop
+    }
```
Comment (kooriookami) [inline packages/components/anchor/src/helper.ts:95]: If any function is useful in vueuse, we can use them.
https://vueuse.org/core/useWindowScroll/
https://vueuse.org/core/useScroll/

File: docs/en-US/component/anchor.md (hunk 12 above)
Comment (Fuphoenixes) [reply docs/en-US/component/anchor.md]: done

File: packages/components/anchor/src/helper.ts (hunk 14 above)
Comment (Fuphoenixes) [reply packages/components/anchor/src/helper.ts:95]: After using the useScroll method for modification, the scrolling animation will be lost in the docs.
[there the code](https://github.com/Fuphoenixes/element-plus/tree/anchor-useScroll)

**Review by jw-foss — COMMENTED** [review]
Hi @Fuphoenixes I left quite a lot comments for your consideration, if you have any questions for me please let me know! Most of the code LGTM, but I do have some questions about the naming and structuring. 👍

**Comment by Fuphoenixes** [comment]
> Hi @Fuphoenixes I left quite a lot comments for your consideration, if you have any questions for me please let me know! Most of the code LGTM, but I do have some questions about the naming and structuring. 👍

Thank you for your review, LGTM, I have changed it !

File: packages/components/anchor/src/anchor.vue [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+  if (!containerEle.value) return
+  isScrolling = true
+  const target = getElement(href) as HTMLElement
+  if (!target) return
+  const scrollEle = getScrollElement(target, containerEle.value)
+  const distance = getOffsetTopDistance(target, scrollEle)
+  const max = scrollEle.scrollHeight - scrollEle.clientHeight
+  const to = Math.min(distance - props.offset, max)
+  animateScrollTo(containerEle.value, preScrollTop, to, props.duration, () => {
+    setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (Fuphoenixes) [reply packages/components/anchor/src/anchor.vue]: I want make sure it is executed after throttleByRaf's handleScroll, but `nextTick` can't do it, so I set this 20ms timeout.

**Review by kooriookami — APPROVED** [review]

**Pull request merged**